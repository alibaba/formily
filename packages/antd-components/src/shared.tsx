import React from 'react'
import { mapTextComponent, mapStyledProps, normalizeCol } from '@formily/antd'
import { Select as AntSelect } from 'antd'
import { SelectProps as AntSelectProps } from 'antd/lib/select'
import styled from 'styled-components'
import { isArr, FormPath } from '@formily/shared'
export * from '@formily/shared'

export const compose = (...args: any[]) => {
  return (payload: any, ...extra: any[]) => {
    return args.reduce((buf, fn) => {
      return buf !== undefined ? fn(buf, ...extra) : fn(payload, ...extra)
    }, payload)
  }
}

interface SelectOption {
  label: React.ReactText
  value: any
  [key: string]: any
}

type SelectProps = AntSelectProps & {
  dataSource?: SelectOption[]
}

const createEnum = (enums: any) => {
  if (isArr(enums)) {
    return enums.map(item => {
      if (typeof item === 'object') {
        return {
          ...item
        }
      } else {
        return {
          label: item,
          value: item
        }
      }
    })
  }

  return []
}

export const Select: React.FC<SelectProps> = styled((props: SelectProps) => {
  const { dataSource = [], onChange, ...others } = props
  const children = createEnum(dataSource).map(item => {
    const { label, value, ...others } = item
    return (
      <AntSelect.Option
        key={value}
        {...others}
        title={label as string}
        value={value}
      >
        {label}
      </AntSelect.Option>
    )
  })
  return (
    <AntSelect
      className={props.className}
      {...others}
      onChange={(value: any, options: any) => {
        onChange(
          value,
          isArr(options)
            ? options.map(item => ({
                ...item,
                props: undefined
              }))
            : {
                ...options,
                props: undefined //干掉循环引用
              }
        )
      }}
    >
      {children}
    </AntSelect>
  )
})`
  min-width: 100px;
  width: 100%;
`
export const acceptEnum = (component: React.JSXElementConstructor<any>) => {
  return ({ dataSource, ...others }) => {
    if (dataSource) {
      return React.createElement(Select, { dataSource, ...others })
    } else {
      return React.createElement(component, others)
    }
  }
}

export const transformDataSourceKey = (component, dataSourceKey) => {
  return ({ dataSource, ...others }) => {
    return React.createElement(component, {
      [dataSourceKey]: dataSource,
      ...others
    })
  }
}

export const createMatchUpdate = (name: string, path: string) => (
  targetName: string,
  targetPath: string,
  callback: () => void
) => {
  if (targetName || targetPath) {
    if (targetName) {
      if (FormPath.parse(targetName).matchAliasGroup(name, path)) {
        callback()
      }
    } else if (targetPath) {
      if (FormPath.parse(targetPath).matchAliasGroup(name, path)) {
        callback()
      }
    }
  } else {
    callback()
  }
}

export { mapTextComponent, mapStyledProps, normalizeCol }
