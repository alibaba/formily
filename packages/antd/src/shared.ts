import React from 'react'
import { PreviewText } from '@formily/react-shared-components'
import {
  IConnectProps,
  MergedFieldComponentProps
} from '@formily/react-schema-renderer'
export * from '@formily/shared'

export const autoScrollInValidateFailed = (formRef: any) => {
  if (formRef.current) {
    setTimeout(() => {
      const elements = formRef.current.querySelectorAll(
        '.next-form-item.has-error'
      )
      if (elements && elements.length) {
        if (!elements[0].scrollIntoView) return
        elements[0].scrollIntoView({
          behavior: 'smooth',
          inline: 'center',
          block: 'center'
        })
      }
    }, 30)
  }
}

export const mapTextComponent = (
  Target: React.JSXElementConstructor<any>,
  props: any,
  fieldProps: any = {}
): React.JSXElementConstructor<any> => {
  const { editable } = fieldProps
  if (editable !== undefined) {
    if (editable === false) {
      return PreviewText
    }
  }
  return Target
}

export const transformDataSourceKey = (component, dataSourceKey) => {
  return ({ dataSource, ...others }) => {
    return React.createElement(component, {
      [dataSourceKey]: dataSource,
      ...others
    })
  }
}

export const normalizeCol = (
  col: { span: number; offset?: number } | number,
  defaultValue?: { span: number }
): { span: number; offset?: number } => {
  if (!col) {
    return defaultValue
  } else {
    return typeof col === 'object' ? col : { span: Number(col) }
  }
}

export const mapStyledProps = (
  props: IConnectProps,
  fieldProps: MergedFieldComponentProps
) => {
  const { loading, errors } = fieldProps
  if (loading) {
    props.state = props.state || 'loading'
  } else if (errors && errors.length) {
    props.state = 'error'
  }
}

export const compose = (...args: any[]) => {
  return (payload: any, ...extra: any[]) => {
    return args.reduce((buf, fn) => {
      return buf !== undefined ? fn(buf, ...extra) : fn(payload, ...extra)
    }, payload)
  }
}
