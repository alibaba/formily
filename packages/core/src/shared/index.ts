
export * from './constants'
export * from './externals'
export * from './internals'
export * from './effectbox'