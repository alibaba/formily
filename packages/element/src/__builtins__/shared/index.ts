export * from './transform-component'
export * from './resolve-component'
export * from './create-context'
export * from './utils'
export * from './portal'
export * from './loading'
export * from './types'
