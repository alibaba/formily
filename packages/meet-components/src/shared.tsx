import React from 'react'
import {
  mapTextComponent,
  mapStyledProps,
  normalizeCol,
  mapSelectComponent
} from '@formily/meet'
import { Select as MeetSelect } from '@alifd/meet'
import { SelectComponent as MeetSelectProps } from '@alifd/meet/types/select'
import styled from 'styled-components'
export * from '@formily/shared'

export const compose = (...args: any[]) => {
  return (payload: any, ...extra: any[]) => {
    return args.reduce((buf, fn) => {
      return buf !== undefined ? fn(buf, ...extra) : fn(payload, ...extra)
    }, payload)
  }
}

interface SelectOption {
  label: React.ReactText
  value: any
  [key: string]: any
}

type SelectProps = MeetSelectProps & {
  dataSource?: SelectOption[]
}

const Select: React.FC<SelectProps> = styled((props: SelectProps) => {
  const { dataSource = [], ...others } = props
  const children = dataSource.map(item => {
    const { label, value, ...others } = item
    return (
      <MeetSelect key={value} {...others} title={label as string} value={value}>
        {label}
      </MeetSelect>
    )
  })
  return (
    <MeetSelect className={props.className} {...others}>
      {children}
    </MeetSelect>
  )
})`
  min-width: 100px;
  width: 100%;
`
export const acceptEnum = (component: React.JSXElementConstructor<any>) => {
  return ({ dataSource, ...others }) => {
    if (dataSource) {
      return React.createElement(Select, { dataSource, ...others })
    } else {
      return React.createElement(component, others)
    }
  }
}

export const transformDataSourceKey = (component, dataSourceKey) => {
  return ({ dataSource, ...others }) => {
    return React.createElement(component, {
      [dataSourceKey]: dataSource,
      ...others
    })
  }
}

export { mapTextComponent, mapStyledProps, normalizeCol, mapSelectComponent }
