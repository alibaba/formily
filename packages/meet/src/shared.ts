import React from 'react'
import { PreviewText } from '@formily/react-shared-components'
import {
  IConnectProps,
  MergedFieldComponentProps
} from '@formily/react-schema-renderer'
import { Select } from '@alifd/meet'
export * from '@formily/shared'

export const mapSelectComponent = (
  Target: React.JSXElementConstructor<any>
  // props: any = {},
  // fieldProps: any = {}
): React.JSXElementConstructor<any> => {
  // console.log(props, fieldProps)
  // const { picker } = props
  // if (picker === 'radio') {
  //   return RadioSelect
  // }
  return Target
}

export const normalizeCol = (
  col: { span: number; offset?: number } | number,
  defaultValue?: { span: number }
): { span: number; offset?: number } => {
  if (!col) {
    return defaultValue
  } else {
    return typeof col === 'object' ? col : { span: Number(col) }
  }
}

export const mapTextComponent = (
  Target: React.JSXElementConstructor<any>,
  props: any = {},
  fieldProps: any = {}
): React.JSXElementConstructor<any> => {
  const { editable } = fieldProps
  if (editable !== undefined) {
    if (editable === false) {
      return PreviewText
    }
  }
  if (Array.isArray(props.dataSource)) {
    return Select
  }
  return Target
}

export const acceptEnum = (component: React.JSXElementConstructor<any>) => {
  return ({ dataSource, ...others }) => {
    if (dataSource) {
      return React.createElement(Select, { dataSource, ...others })
    } else {
      return React.createElement(component, others)
    }
  }
}

export const transformDataSourceKey = (component, dataSourceKey) => {
  return ({ dataSource, ...others }) => {
    return React.createElement(component, {
      [dataSourceKey]: dataSource,
      ...others
    })
  }
}

export const mapStyledProps = (
  props: IConnectProps,
  fieldProps: MergedFieldComponentProps
) => {
  const { loading, errors } = fieldProps
  if (loading) {
    props.state = props.state || 'loading'
  } else if (errors && errors.length) {
    props.state = 'error'
  }
}

export const compose = (...args: any[]) => {
  return (payload: any, ...extra: any[]) => {
    return args.reduce((buf, fn) => {
      return buf !== undefined ? fn(buf, ...extra) : fn(payload, ...extra)
    }, payload)
  }
}
