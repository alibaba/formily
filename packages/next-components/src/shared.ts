import React from 'react'
import { mapTextComponent, mapStyledProps, normalizeCol } from '@formily/next'
import { Select } from '@alifd/next'
export * from '@formily/shared'

export const compose = (...args: any[]) => {
  return (payload: any, ...extra: any[]) => {
    return args.reduce((buf, fn) => {
      return buf !== undefined ? fn(buf, ...extra) : fn(payload, ...extra)
    }, payload)
  }
}

export const acceptEnum = (component: React.JSXElementConstructor<any>) => {
  return ({ dataSource, ...others }) => {
    if (dataSource) {
      return React.createElement(Select, { dataSource, ...others })
    } else {
      return React.createElement(component, others)
    }
  }
}

export { mapTextComponent, mapStyledProps, normalizeCol }
