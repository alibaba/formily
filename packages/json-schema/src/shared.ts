import { isFn, each, isPlainObj, isArr, toArr, FormPath } from '@formily/shared'
import { isObservable, untracked } from '@formily/reactive'
import { Schema } from './schema'
import { ISchema } from './types'

const REVA_ACTIONS_KEY = Symbol.for('__REVA_ACTIONS')

export const SchemaNestedMap = {
  parent: true,
  root: true,
  properties: true,
  patternProperties: true,
  additionalProperties: true,
  items: true,
  additionalItems: true,
  'x-linkages': true,
  'x-reactions': true,
}

export const SchemaStateMap = {
  title: 'title',
  description: 'description',
  default: 'initialValue',
  enum: 'dataSource',
  readOnly: 'readOnly',
  writeOnly: 'editable',
  'x-content': 'content',
  'x-data': 'data',
  'x-value': 'value',
  'x-editable': 'editable',
  'x-disabled': 'disabled',
  'x-read-pretty': 'readPretty',
  'x-read-only': 'readOnly',
  'x-visible': 'visible',
  'x-hidden': 'hidden',
  'x-display': 'display',
  'x-pattern': 'pattern',
  'x-validator': 'validator',
  'x-decorator': 'decoratorType',
  'x-component': 'componentType',
  'x-decorator-props': 'decoratorProps',
  'x-component-props': 'componentProps',
}

export const SchemaValidatorMap = {
  required: true,
  format: true,
  maxItems: true,
  minItems: true,
  maxLength: true,
  minLength: true,
  maximum: true,
  minimum: true,
  exclusiveMaximum: true,
  exclusiveMinimum: true,
  pattern: true,
  const: true,
  multipleOf: true,
  maxProperties: true,
  minProperties: true,
  uniqueItems: true,
}

export const SchemaNormalKeys = Object.keys(SchemaStateMap)

export const SchemaValidatorKeys = Object.keys(SchemaValidatorMap)

export const hasOwnProperty = Object.prototype.hasOwnProperty

export const traverse = (
  target: any,
  visitor: (value: any, path: Array<string | number>) => void
) => {
  const seenObjects = []
  const root = target
  const traverse = (target: any, path = []) => {
    if (isPlainObj(target)) {
      const seenIndex = seenObjects.indexOf(target)
      if (seenIndex > -1) {
        return
      }
      const addIndex = seenObjects.length
      seenObjects.push(target)
      if (isNoNeedCompileObject(target) && root !== target) {
        visitor(target, path)
        return
      }
      each(target, (value, key) => {
        traverse(value, path.concat(key))
      })
      seenObjects.splice(addIndex, 1)
    } else {
      visitor(target, path)
    }
  }
  traverse(target)
}

export const traverseSchema = (
  schema: ISchema,
  visitor: (value: any, path: any[], omitCompile?: boolean) => void
) => {
  if (schema['x-validator'] !== undefined) {
    visitor(schema['x-validator'], ['x-validator'])
  }
  const seenObjects = []
  const root = schema
  const traverse = (target: any, path = []) => {
    if (
      path[0] === 'x-compile-omitted' ||
      path[0] === 'x-validator' ||
      path[0] === 'version' ||
      path[0] === '_isJSONSchemaObject'
    )
      return
    if (String(path[0]).indexOf('x-') == -1 && isFn(target)) return
    if (SchemaNestedMap[path[0]]) return
    if (schema['x-compile-omitted']?.indexOf(path[0]) > -1) {
      visitor(target, path, true)
      return
    }
    if (isPlainObj(target)) {
      if (path[0] === 'default' || path[0] === 'x-value') {
        visitor(target, path)
        return
      }
      const seenIndex = seenObjects.indexOf(target)
      if (seenIndex > -1) {
        return
      }
      const addIndex = seenObjects.length
      seenObjects.push(target)
      if (isNoNeedCompileObject(target) && root !== target) {
        visitor(target, path)
        return
      }
      each(target, (value, key) => {
        traverse(value, path.concat(key))
      })
      seenObjects.splice(addIndex, 1)
    } else {
      visitor(target, path)
    }
  }
  traverse(schema)
}

export const isNoNeedCompileObject = (source: any) => {
  if ('$$typeof' in source && '_owner' in source) {
    return true
  }
  if (source['_isAMomentObject']) {
    return true
  }
  if (Schema.isSchemaInstance(source)) {
    return true
  }
  if (source[REVA_ACTIONS_KEY]) {
    return true
  }
  if (isFn(source['toJS'])) {
    return true
  }
  if (isFn(source['toJSON'])) {
    return true
  }
  if (isObservable(source)) {
    return true
  }
  return false
}

export const createDataSource = (source: any[]) => {
  return toArr(source).map((item) => {
    if (typeof item === 'object') {
      return item
    } else {
      return {
        label: item,
        value: item,
      }
    }
  })
}

export const patchStateFormSchema = (
  targetState: any,
  pattern: any[],
  compiled: any
) => {
  untracked(() => {
    const path = FormPath.parse(pattern)
    const segments = path.segments
    const key = segments[0]
    const isEnum = key === 'enum' && isArr(compiled)
    const schemaMapKey = SchemaStateMap[key]
    if (schemaMapKey) {
      FormPath.setIn(
        targetState,
        [schemaMapKey].concat(segments.slice(1)),
        isEnum ? createDataSource(compiled) : compiled
      )
    } else {
      const isValidatorKey = SchemaValidatorMap[key]
      if (isValidatorKey) {
        targetState['setValidatorRule']?.(key, compiled)
      }
    }
  })
}
