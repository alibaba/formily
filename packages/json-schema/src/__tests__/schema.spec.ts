import { Schema } from '../'
import { isFn } from '@formily/shared'

test('has methods', () => {
  const schema = new Schema({
    type: 'object',
    properties: {
      aa: {
        type: 'string',
      },
    },
  })
  expect(isFn(schema.setAdditionalItems)).toBeTruthy()
  expect(isFn(schema.setAdditionalProperties)).toBeTruthy()
  expect(isFn(schema.setItems)).toBeTruthy()
  expect(isFn(schema.setPatternProperties)).toBeTruthy()
  expect(isFn(schema.setProperties)).toBeTruthy()
  expect(isFn(schema.addPatternProperty)).toBeTruthy()
  expect(isFn(schema.addProperty)).toBeTruthy()
  expect(isFn(schema.fromJSON)).toBeTruthy()
  expect(isFn(schema.toJSON)).toBeTruthy()
  expect(isFn(schema.reducePatternProperties)).toBeTruthy()
  expect(isFn(schema.reduceProperties)).toBeTruthy()
  expect(isFn(schema.removeProperty)).toBeTruthy()
  expect(isFn(schema.removePatternProperty)).toBeTruthy()
  expect(isFn(schema.mapPatternProperties)).toBeTruthy()
  expect(isFn(schema.mapProperties)).toBeTruthy()

  expect(isFn(Schema.isSchemaInstance)).toBeTruthy()
  expect(isFn(Schema.registerCompiler)).toBeTruthy()
  expect(isFn(Schema.registerPatches)).toBeTruthy()
  expect(isFn(Schema.shallowCompile)).toBeTruthy()
  expect(isFn(Schema.compile)).toBeTruthy()
  expect(isFn(Schema.getOrderProperties)).toBeTruthy()
})

test('all props', () => {
  const schema = new Schema({
    type: 'object',
    title: 'title',
    description: 'description',
    patternProperties: {
      '^[a-zA-Z0-9]*$': {
        properties: {
          model: { type: 'string' },
          made: { type: 'string' },
          year: { type: 'string' },
        },
      },
    },
    additionalProperties: {
      type: 'string',
    },
    properties: {
      string: {
        type: 'string',
        default: 'default',
        required: true,
        'x-component': 'Input',
        'x-component-props': {
          placeholder: 'placeholder',
        },
        'x-decorator': 'FormItem',
        'x-decorator-props': {
          labelCol: 3,
        },
        'x-disabled': true,
        'x-display': 'visible',
        'x-editable': false,
        'x-hidden': false,
        'x-pattern': 'readPretty',
        'x-read-only': true,
        'x-validator': ['phone'],
        'x-reactions': [
          {
            target: 'xxx',
            when: '{{aa > bb}}',
          },
        ],
      },
      boolean: {
        type: 'boolean',
        default: false,
      },
      number: {
        type: 'number',
        default: 100,
      },
      date: {
        type: 'date',
        default: '2020-12-23',
      },
      datetime: {
        type: 'datetime',
        default: '2020-12-23 23:00:00',
      },
      array: {
        type: 'array',
        items: {
          type: 'string',
        },
        additionalItems: {
          type: 'number',
        },
      },
      array2: {
        type: 'array',
        items: [
          {
            type: 'string',
          },
          {
            type: 'object',
          },
        ],
      },
      void: {
        type: 'void',
      },
    },
  })
  expect(schema).toMatchSnapshot()
})

test('all methods', () => {
  const schema = new Schema({
    type: 'object',
    'x-reactions': null,
  })
  const schema2 = new Schema({
    type: 'object',
    fn: () => {},
  } as any)
  const schema3 = new Schema({
    type: 'object',
    additionalItems: null,
    additionalProperties: null,
    properties: null,
  })
  schema3.additionalItems = null
  schema3.additionalProperties = null
  schema3.properties = {
    xxx: null,
  }
  schema3.items = [null]
  const schema4 = new Schema({
    type: 'object',
    additionalItems: {},
    additionalProperties: {},
    properties: {},
  })
  schema4.additionalItems = {} as any
  schema4.additionalProperties = {} as any
  schema4.properties = {
    xxx: {} as any,
  }
  schema4.items = [{}] as any
  const schema5 = new Schema({
    type: 'array',
  })
  schema5.items = null
  const schema6 = new Schema({
    type: 'array',
  })
  schema6.items = {} as any
  const schema7 = new Schema({
    type: 'array',
    items: {
      type: 'string',
    },
  })
  const string = schema.addProperty('string', {
    type: 'string',
    title: 'string',
    description: null,
    'x-reactions': [
      {
        target: 'xxx',
        when: true,
        fulfill: {
          schema: {},
        },
      },
    ],
  })

  const array = schema.addProperty('array', {
    type: 'string',
    title: 'string',
    items: [{ type: 'integer' }, { type: 'integer' }],
  })
  const pattern = schema.addPatternProperty('^[a-zA-Z0-9]*$', {
    properties: {
      model: { type: 'string', 'x-index': 2 },
      made: { type: 'string', 'x-index': 1 },
      year: { type: 'string', 'x-index': 0 },
    },
  })
  schema.addPatternProperty('xxx', null)
  schema.setAdditionalProperties({
    type: 'string',
  })
  schema.setAdditionalProperties(null)
  array.setItems(null)
  array.setAdditionalItems({
    type: 'string',
  })
  array.setAdditionalItems(null)
  schema.setPatternProperties(null)
  schema.fromJSON(null)
  expect(schema2['fn']).toBeUndefined()
  expect(schema.properties.string).not.toBeUndefined()
  expect(schema.patternProperties['^[a-zA-Z0-9]*$']).not.toBeUndefined()
  expect(schema).toMatchSnapshot()
  expect(schema.toJSON()).toMatchSnapshot()
  expect(pattern.mapProperties((schema, key) => key)).toEqual([
    'year',
    'made',
    'model',
  ])
  expect(
    pattern.reduceProperties((buf, schema, key) => buf.concat('_' + key), [])
  ).toEqual(['_year', '_made', '_model'])
  expect(schema.mapPatternProperties((schema, key) => key)).toEqual([
    '^[a-zA-Z0-9]*$',
  ])
  expect(
    schema.reducePatternProperties(
      (buf, schema, key) => buf.concat('_' + key),
      []
    )
  ).toEqual(['_^[a-zA-Z0-9]*$'])
  schema5.toJSON()
  schema6.toJSON()
  schema7.toJSON()
  schema.removeProperty('string')
  expect(schema.properties.string).toBeUndefined()
  schema.removePatternProperty('^[a-zA-Z0-9]*$')
  expect(schema.patternProperties['^[a-zA-Z0-9]*$']).toBeUndefined()
  expect(schema.compile()).toMatchSnapshot()
  expect(string.compile()).toMatchSnapshot()
  expect(schema3.toJSON()).toMatchSnapshot()
  expect(schema4.toJSON()).toMatchSnapshot()
})

describe('all static methods', () => {
  expect(Schema.compile({ aa: '{{123}}' })).toEqual({ aa: 123 })
  expect(Schema.shallowCompile('{{123}}')).toEqual(123)
  expect(Schema.getOrderProperties()).toEqual([])
  Schema.registerPatches(null)
})

test('single function x-reactions', () => {
  const reactions = () => console.info('x-reactions')
  const schema = new Schema({
    type: 'string',
    'x-reactions': reactions,
  })
  expect(schema.compile()['x-reactions']).toEqual(reactions)
})

test('definitions and $ref', () => {
  const schema = new Schema({
    definitions: {
      address: {
        type: 'object',
        properties: {
          street_address: {
            type: 'string',
          },
          city: {
            type: 'string',
          },
          state: {
            type: 'string',
          },
        },
        required: ['street_address', 'city', 'state'],
      },
    },
    type: 'object',
    properties: {
      billing_address: {
        title: 'Billing address',
        $ref: '#/definitions/address',
      },
      shipping_address: {
        title: 'Shipping address',
        $ref: '#/definitions/address',
      },
    },
  })
  expect(schema.properties.billing_address.required).toEqual([
    'street_address',
    'city',
    'state',
  ])
})
