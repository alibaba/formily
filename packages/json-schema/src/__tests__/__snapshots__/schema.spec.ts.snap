// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all methods 1`] = `
Object {
  "_isJSONSchemaObject": true,
  "additionalProperties": Object {
    "_isJSONSchemaObject": true,
    "type": "string",
    "version": "2.0",
  },
  "patternProperties": Object {
    "^[a-zA-Z0-9]*$": Object {
      "_isJSONSchemaObject": true,
      "name": "^[a-zA-Z0-9]*$",
      "properties": Object {
        "made": Object {
          "_isJSONSchemaObject": true,
          "name": "made",
          "type": "string",
          "version": "2.0",
          "x-index": 1,
        },
        "model": Object {
          "_isJSONSchemaObject": true,
          "name": "model",
          "type": "string",
          "version": "2.0",
          "x-index": 2,
        },
        "year": Object {
          "_isJSONSchemaObject": true,
          "name": "year",
          "type": "string",
          "version": "2.0",
          "x-index": 0,
        },
      },
      "version": "2.0",
    },
  },
  "properties": Object {
    "array": Object {
      "_isJSONSchemaObject": true,
      "additionalItems": Object {
        "_isJSONSchemaObject": true,
        "type": "string",
        "version": "2.0",
      },
      "items": Array [
        Object {
          "_isJSONSchemaObject": true,
          "type": "integer",
          "version": "2.0",
        },
        Object {
          "_isJSONSchemaObject": true,
          "type": "integer",
          "version": "2.0",
        },
      ],
      "name": "array",
      "title": "string",
      "type": "string",
      "version": "2.0",
    },
    "string": Object {
      "_isJSONSchemaObject": true,
      "description": null,
      "name": "string",
      "title": "string",
      "type": "string",
      "version": "2.0",
      "x-reactions": Array [
        Object {
          "fulfill": Object {
            "schema": Object {},
          },
          "target": "xxx",
          "when": true,
        },
      ],
    },
  },
  "type": "object",
  "version": "2.0",
  "x-reactions": null,
}
`;

exports[`all methods 2`] = `
Object {
  "_isJSONSchemaObject": true,
  "additionalProperties": Object {
    "_isJSONSchemaObject": true,
    "type": "string",
    "version": "2.0",
  },
  "patternProperties": Object {
    "^[a-zA-Z0-9]*$": Object {
      "_isJSONSchemaObject": true,
      "name": "^[a-zA-Z0-9]*$",
      "properties": Object {
        "made": Object {
          "_isJSONSchemaObject": true,
          "name": "made",
          "type": "string",
          "version": "2.0",
          "x-index": 1,
        },
        "model": Object {
          "_isJSONSchemaObject": true,
          "name": "model",
          "type": "string",
          "version": "2.0",
          "x-index": 2,
        },
        "year": Object {
          "_isJSONSchemaObject": true,
          "name": "year",
          "type": "string",
          "version": "2.0",
          "x-index": 0,
        },
      },
      "version": "2.0",
    },
  },
  "properties": Object {
    "array": Object {
      "_isJSONSchemaObject": true,
      "additionalItems": Object {
        "_isJSONSchemaObject": true,
        "type": "string",
        "version": "2.0",
      },
      "items": Array [
        Object {
          "_isJSONSchemaObject": true,
          "type": "integer",
          "version": "2.0",
        },
        Object {
          "_isJSONSchemaObject": true,
          "type": "integer",
          "version": "2.0",
        },
      ],
      "name": "array",
      "title": "string",
      "type": "string",
      "version": "2.0",
    },
    "string": Object {
      "_isJSONSchemaObject": true,
      "description": null,
      "name": "string",
      "title": "string",
      "type": "string",
      "version": "2.0",
      "x-reactions": Array [
        Object {
          "fulfill": Object {
            "schema": Object {},
          },
          "target": "xxx",
          "when": true,
        },
      ],
    },
  },
  "type": "object",
  "version": "2.0",
  "x-reactions": null,
}
`;

exports[`all methods 3`] = `
Object {
  "_isJSONSchemaObject": true,
  "additionalProperties": Object {
    "_isJSONSchemaObject": true,
    "type": "string",
    "version": "2.0",
  },
  "patternProperties": Object {},
  "properties": Object {
    "array": Object {
      "_isJSONSchemaObject": true,
      "additionalItems": Object {
        "_isJSONSchemaObject": true,
        "type": "string",
        "version": "2.0",
      },
      "items": Array [
        Object {
          "_isJSONSchemaObject": true,
          "type": "integer",
          "version": "2.0",
        },
        Object {
          "_isJSONSchemaObject": true,
          "type": "integer",
          "version": "2.0",
        },
      ],
      "name": "array",
      "title": "string",
      "type": "string",
      "version": "2.0",
    },
  },
  "type": "object",
  "version": "2.0",
  "x-reactions": null,
}
`;

exports[`all methods 4`] = `
Object {
  "_isJSONSchemaObject": true,
  "description": null,
  "name": "string",
  "title": "string",
  "type": "string",
  "version": "2.0",
  "x-reactions": Array [
    Object {
      "fulfill": Object {
        "schema": Object {},
      },
      "target": "xxx",
      "when": true,
    },
  ],
}
`;

exports[`all methods 5`] = `
Object {
  "_isJSONSchemaObject": true,
  "additionalItems": undefined,
  "additionalProperties": undefined,
  "items": Array [
    undefined,
  ],
  "properties": Object {
    "xxx": undefined,
  },
  "type": "object",
  "version": "2.0",
}
`;

exports[`all methods 6`] = `
Object {
  "_isJSONSchemaObject": true,
  "additionalItems": undefined,
  "additionalProperties": undefined,
  "items": Array [
    undefined,
  ],
  "properties": Object {
    "xxx": undefined,
  },
  "type": "object",
  "version": "2.0",
}
`;

exports[`all props 1`] = `
Object {
  "_isJSONSchemaObject": true,
  "additionalProperties": Object {
    "_isJSONSchemaObject": true,
    "type": "string",
    "version": "2.0",
  },
  "description": "description",
  "patternProperties": Object {
    "^[a-zA-Z0-9]*$": Object {
      "_isJSONSchemaObject": true,
      "name": "^[a-zA-Z0-9]*$",
      "properties": Object {
        "made": Object {
          "_isJSONSchemaObject": true,
          "name": "made",
          "type": "string",
          "version": "2.0",
        },
        "model": Object {
          "_isJSONSchemaObject": true,
          "name": "model",
          "type": "string",
          "version": "2.0",
        },
        "year": Object {
          "_isJSONSchemaObject": true,
          "name": "year",
          "type": "string",
          "version": "2.0",
        },
      },
      "version": "2.0",
    },
  },
  "properties": Object {
    "array": Object {
      "_isJSONSchemaObject": true,
      "additionalItems": Object {
        "_isJSONSchemaObject": true,
        "type": "number",
        "version": "2.0",
      },
      "items": Object {
        "_isJSONSchemaObject": true,
        "type": "string",
        "version": "2.0",
      },
      "name": "array",
      "type": "array",
      "version": "2.0",
    },
    "array2": Object {
      "_isJSONSchemaObject": true,
      "items": Array [
        Object {
          "_isJSONSchemaObject": true,
          "type": "string",
          "version": "2.0",
        },
        Object {
          "_isJSONSchemaObject": true,
          "type": "object",
          "version": "2.0",
        },
      ],
      "name": "array2",
      "type": "array",
      "version": "2.0",
    },
    "boolean": Object {
      "_isJSONSchemaObject": true,
      "default": false,
      "name": "boolean",
      "type": "boolean",
      "version": "2.0",
    },
    "date": Object {
      "_isJSONSchemaObject": true,
      "default": "2020-12-23",
      "name": "date",
      "type": "date",
      "version": "2.0",
    },
    "datetime": Object {
      "_isJSONSchemaObject": true,
      "default": "2020-12-23 23:00:00",
      "name": "datetime",
      "type": "datetime",
      "version": "2.0",
    },
    "number": Object {
      "_isJSONSchemaObject": true,
      "default": 100,
      "name": "number",
      "type": "number",
      "version": "2.0",
    },
    "string": Object {
      "_isJSONSchemaObject": true,
      "default": "default",
      "name": "string",
      "required": true,
      "type": "string",
      "version": "2.0",
      "x-component": "Input",
      "x-component-props": Object {
        "placeholder": "placeholder",
      },
      "x-decorator": "FormItem",
      "x-decorator-props": Object {
        "labelCol": 3,
      },
      "x-disabled": true,
      "x-display": "visible",
      "x-editable": false,
      "x-hidden": false,
      "x-pattern": "readPretty",
      "x-reactions": Array [
        Object {
          "target": "xxx",
          "when": "{{aa > bb}}",
        },
      ],
      "x-read-only": true,
      "x-validator": Array [
        "phone",
      ],
    },
    "void": Object {
      "_isJSONSchemaObject": true,
      "name": "void",
      "type": "void",
      "version": "2.0",
    },
  },
  "title": "title",
  "type": "object",
  "version": "2.0",
}
`;
